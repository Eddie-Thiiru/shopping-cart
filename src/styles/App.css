#App {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-rows: 100px minmax(800px, 1fr) 400px;
}

.header {
  position: sticky;
  width: 100%;
  top: 0;
  /* background-color: rgba(0, 0, 0, 0.8); */
  background: linear-gradient(to bottom, rgb(0, 0, 0), rgba(0, 0, 0, 0.8));
  padding: 10px;
  display: grid;
  grid-template-columns: 1fr 2fr;
  align-items: center;
  gap: 20px;
  z-index: 1;
}

.mainSection {
  padding: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.footer {
  width: 100%;
  padding: 10px;
  display: grid;
  grid-template-rows: repeat(3, 1fr);
  justify-content: center;
  align-items: center;
}

.nav {
  padding: 10px;
  display: flex;
  gap: 10px;
}

.nav > div {
  margin-left: auto;
}

.footer > div {
  display: flex;
  flex-direction: column;
  justify-items: center;
  align-items: center;
  gap: 20px;
}

.footerAbout,
.footerSocial,
.footerPolicies {
  display: flex;
  gap: 30px;
}

.footerCopyright {
  display: flex;
  flex-direction: column;
  gap: 5px;
}

.apiHyperlink img {
  height: 50px;
}

.apiHyperlink a {
  display: flex;
  align-items: center;
  gap: 10px;
}

ul {
  list-style: none;
}

h1 {
  font-size: 2.5em;
  line-height: 1.1;
}

img {
  object-fit: cover;
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}
a:hover {
  color: #535bf2;
}
