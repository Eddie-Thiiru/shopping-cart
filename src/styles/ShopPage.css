.shopPage {
  height: 100%;
  display: grid;
  grid-template-columns: 100px 1fr 100px;
  grid-template-rows: 50px 1fr;
}

.shopPage > div {
  grid-column: 2 / 3;
}

.emptyCart {
  grid-row: 2 / 3;
}

.cart {
  display: grid;
  grid-template-rows: 50px 1fr 150px;
}

.cartItemCard {
  height: 100px;
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.cartItemWrapperOne {
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.cartItemWrapperOne img {
  max-width: 50%;
}
