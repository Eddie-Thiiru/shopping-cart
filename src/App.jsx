import { useState, createContext } from "react";
import Header from "./components/Header";
import MainSection from "./components/MainSection";
import Footer from "./components/Footer";
import { DataProvider } from "./components/DataProvider";
import "./styles/App.css";

export const CartContext = createContext();

const App = () => {
  const [cartItems, setCartItems] = useState({
    count: 0,
    items: [],
  });

  const isDuplicate = (id) => {
    const array = cartItems.items;

    for (let i = 0; i < array.length; i++) {
      const element = array[i];

      if (element.uniqueId === id) {
        return true;
      }
    }
    return false;
  };

  const addToCart = (data) => {
    const duplicate = isDuplicate(data.uniqueId);

    if (duplicate === false) {
      setCartItems({
        count: cartItems.count + 1,
        items: cartItems.items.concat(data),
      });
    } else {
      return;
    }
  };

  const removeFromCart = (id) => {
    setCartItems({
      count: cartItems.count - 1,
      items: cartItems.items.filter((item) => item.uniqueId !== id),
    });
  };

  const cartData = cartItems.items;

  return (
    <div id="App">
      <Header cartItemsCount={cartItems.count} />
      <DataProvider>
        <CartContext.Provider value={{ cartData, addToCart, removeFromCart }}>
          <MainSection />
        </CartContext.Provider>
      </DataProvider>
      <Footer />
    </div>
  );
};

export default App;
