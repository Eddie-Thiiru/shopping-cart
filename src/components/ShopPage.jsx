import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import PropTypes from "prop-types";
import { CartContext } from "../App";
import { RiDeleteBinLine } from "react-icons/ri";
import "../styles/ShopPage.css";

const CartCard = ({ imageURL, title, price, id, handleDelete }) => {
  return (
    <div className="cartItemCard">
      <div className="cartItemWrapperOne">
        <img src={imageURL} alt={`${title} background image`}></img>
        <p>{title}</p>
      </div>
      <div className="cartItemWrapperTwo">
        <p>{`$${price}`}</p>
        <div>
          <RiDeleteBinLine />
          <button
            type="button"
            className="removeGame"
            onClick={() => handleDelete(id)}
          >
            Remove
          </button>
        </div>
      </div>
    </div>
  );
};

const Empty = () => {
  const navigate = useNavigate();

  const handleClick = () => {
    navigate("/games/collection");
  };

  return (
    <div className="emptyCart">
      <h2>Nothing here</h2>
      <button type="button" className="cartEmptyBtn" onClick={handleClick}>
        Buy Games
      </button>
    </div>
  );
};

const ShopPage = () => {
  const { cartData, removeFromCart } = useContext(CartContext);
  const totalPrice = cartData.reduce(
    (prev, current) => prev + current.price,
    0,
  );

  return (
    <div className="shopPage">
      <div className="pageTitle">
        <h2>Cart</h2>
      </div>
      <div className="cart">
        {cartData.length === 0 ? (
          <Empty />
        ) : (
          <div className="cartItemsContainer">
            <div className="cartHeader">
              <h3>Games</h3>
              <h3>price</h3>
            </div>
            <div className="cartItemsWrapper">
              {cartData.map((item, index) => {
                return (
                  <CartCard
                    key={index}
                    id={item.uniqueId}
                    imageURL={item.imageURL}
                    title={item.name}
                    price={item.price}
                    handleDelete={removeFromCart}
                  />
                );
              })}
            </div>

            <div className="cartTotalWrapper">
              <h3>Grand total</h3>
              <p>{`$${parseFloat(totalPrice).toFixed(2)}`}</p>
              <button type="button" className="paymentBtn">
                Continue to Payment
              </button>
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

CartCard.propTypes = {
  imageURL: PropTypes.string,
  title: PropTypes.string,
  price: PropTypes.number,
  id: PropTypes.number,
  handleDelete: PropTypes.func,
};

export default ShopPage;
