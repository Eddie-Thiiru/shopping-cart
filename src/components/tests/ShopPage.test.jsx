import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ShopPage from "../ShopPage";
import { CartContext } from "../../App";

describe("Shop page", () => {
  it("renders empty cart", () => {
    const cartData = [];

    render(
      <MemoryRouter initialEntries={[{ pathname: "/cart" }]}>
        <CartContext.Provider value={{ cartData }}>
          <ShopPage />
        </CartContext.Provider>
      </MemoryRouter>,
    );

    expect(screen.getByText(/nothing here/i)).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Buy Games" }),
    ).toBeInTheDocument();
  });

  it("should render cart with a product", () => {
    const cartData = [
      {
        imageURL: "https://games-endpoint.example/image",
        name: "Grand Theft Auto 5",
        price: 29.99,
      },
    ];

    render(
      <CartContext.Provider value={{ cartData }}>
        <ShopPage />
      </CartContext.Provider>,
    );

    expect(screen.getByText(/grand Theft Auto 5/i)).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Continue to Payment" }),
    ).toBeInTheDocument();
  });

  // it("should navigate to collection page on empty cart button click", async () => {
  //   const user = userEvent.setup();
  //   const cartData = [];

  //   render(
  //     <BrowserRouter>
  //       <CartContext.Provider value={{ cartData }}>
  //         <ShopPage />
  //       </CartContext.Provider>
  //     </BrowserRouter>,
  //   );

  //   const button = screen.getByRole("button", { name: "Buy Games" });

  //   await user.click(button);

  //   screen.debug();

  //   expect(screen.getByText(/all games/i)).toBeInTheDocument();
  // });
});
