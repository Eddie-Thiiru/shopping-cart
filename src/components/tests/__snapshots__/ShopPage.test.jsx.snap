// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop page > should render shop page component 1`] = `
<div>
  <div
    class="shopPage"
  >
    <div
      class="pageTitle"
    >
      <h2>
        Cart
      </h2>
    </div>
    <div
      class="cart"
    >
      <div
        class="cartHeader"
      >
        <h3>
          Games
        </h3>
        <h3>
          price
        </h3>
      </div>
      <div
        class="cartItemCard"
      >
        <div
          class="cartItemWrapperOne"
        >
          <img
            alt="Grand Theft Auto 5 background image"
            src="https://games-endpoint.example/image"
          />
          <p>
            name name name
          </p>
        </div>
        <div
          class="cartItemWrapperTwo"
        >
          <p>
            29.99
          </p>
          <button
            class="removeGame"
            type="button"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        class="cartTotalWrapper"
      >
        <h3>
          Grand total
        </h3>
        <p>
          29.99
        </p>
        <button
          class="paymentBtn"
          type="button"
        >
          Continue to Payment
        </button>
      </div>
    </div>
  </div>
</div>
`;
